### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by giaur500** [greensopinion/flutter-vector-map-tiles#120]
I noticed performance issues when zooming/moving is actually poor in compare to native solutions like Google Maps or Mapbox official api. On low end devices map is almost unusable.

From other side, rasrer tiles from flutter map plugin are still very fast. Do you aware performance issues and is there anything possible to make it working faster?

**Comment by FaFre** [comment]
You can have a read in issue #10 to get yourself an overview of the situation.

**Comment by FaFre** [comment]
If you are interested, there are some past conversations in the closed issues where you can have a read and might add some ideas. The tile rendering (drawing on canvas) is currently only possible in the flutter main isolate (https://github.com/flutter/flutter/issues/75755), when I'm not mistaken this is still the bottleneck.

You can improve performance by tuning your style. Feel free to do your own experiments and share it with us :)

**Comment by greensopinion** [comment]
We've done a lot of work to address performance issues, but it's still not as good as raster or native maps. Poor performance is most noticeable with tiles having a lot of data (e.g. in LA or London). 
You can find previous discussions by [searching for other issues](https://github.com/greensopinion/flutter-vector-map-tiles/issues?q=is%3Aissue++performance)

The theme can make a big difference as noted by @FaFre (thanks!) but you may find that raster tiles are needed to get the performance that you want, depending on your use-case.

I've added some examples that you may find helpful here: https://github.com/greensopinion/flutter-vector-map-tiles-examples
There are two custom themes in there that may help with performance.

**Comment by giaur500** [comment]
As far as I understand, currently is not possible to optimize it becuse of Flutter limitations? Ok, I will try themes, Indeed, a lot of data causes poor performance.

**Comment by FaFre** [comment]
@greensopinion I pushed the flutter issue and got a response that `flutter_isolate` plugin might improve performance and gain access to dart ui: https://github.com/flutter/flutter/issues/75755#issuecomment-1434935593

However, this plugin is only Android/iOS compatible, do you see any chance this would improve the current performance?

**Comment by greensopinion** [comment]
Yes, there's a good chance that it could be used to offload some CPU-intensive tasks off the UI thread. We are already offloading some work to isolates, but there are many things that can't be offloaded due to use of `dart:ui`.

I imagine that the easiest and most beneficial thing to do would be to render a raster image on an isolate. This could either be integrated directly with the existing `flutter_map` raster tile capabilities, or perhaps in `vector_map_tiles`. The benefit over using raster tiles directly would be choice of themes client-side, and rendering tiles at higher zoom levels that aren't normally available. At 2x (retina) size raster tiles look pretty sharp, but we could even render tiles at incremental zoom levels (e.g. a tile at 14.25 zoom instead of 14 then scaled to 14.25)

I'm a little reluctant to invest in functionality that is using officially unsupported functionality, such as `dart:ui` on non-UI isolates. As noted on [flutter/issues/75755](https://github.com/flutter/flutter/issues/75755#issuecomment-1434935593) things may already break with Impeller.

If we consider moving forward with this idea, we should first evaluate how Impeller affects performance. For example, a comparison running the example library in the profiler with Impeller and without would be really helpful.

**Comment by giaur500** [comment]
@greensopinion do you think it's possible to disable Impeller and use isolates on Android only?

**Comment by greensopinion** [comment]
At least for now it's possible - but that will likely change once Flutter fully adopts Impeller since they won't want to maintain two separate rendering frameworks for Android indefinitely.

**Comment by giaur500** [comment]
I don't know how much work is it, but I think better to use isolates on Android, even as temporary solution. Still more than nothing.

**Comment by greensopinion** [comment]
> However, this plugin is only Android/iOS compatible, do you see any chance this would improve the current performance?

I've pushed an experiment to https://github.com/greensopinion/flutter-vector-map-tiles-raster-experiment
Feel free to try it out!

**Comment by giaur500** [comment]
It works pretty fast, almost the same performance as native Google Maps. Very good result in my opinion. Not tried on iOS though

**Comment by aamirki** [comment]
Hey I just tried out that experiment on iOS and the performance is really good!

**Comment by greensopinion** [comment]
I've created a branch in this repository for experimentation: [flutter-vector-map-tiles/tree/raster-experiment](https://github.com/greensopinion/flutter-vector-map-tiles/tree/raster-experiment)

There are a few things needed to take it further:

1. there may be a memory leak
2. the API is unnatural and needs revisiting
3. there is no way to plug in rendering in a separate isolate
4. tuning

**Comment by Pepslee** [comment]
> > However, this plugin is only Android/iOS compatible, do you see any chance this would improve the current performance?
> 
> I've pushed an experiment to https://github.com/greensopinion/flutter-vector-map-tiles-raster-experiment Feel free to try it out!

It works really cool. But there is some problem in my case. When I'm panning for a long distance it seems that a huge number of isolates creates for all panned tiles. Seems like tile provider doesn't drop deprecated isolates for tiles which was rapidly panned. For example, I have 100 tiles on the map and I`m trying to pan  from 1 tile to 99 tile by one swipe. Tile provider starts to create 99 isolates and tries to process them all but not just skips unnecessary tiles and process only several last tiles.

**Comment by greensopinion** [comment]
The experiment is only intended as a proof-of-concept. There is a lot of work needed to make it ready for real usage.

I've been working on #129 to introduce raster images as an alternative to vectors. Raster images have the advantage of improved frame rate (less jank) and they only need to be rendered the first time since they can be stored in the cache for future use. The result is quite good. Of course, raster images aren't as sharp at fractional zoom levels. #129 is now available as version 3.3.0 of this library.

**Comment by Pepslee** [comment]
> The experiment is only intended as a proof-of-concept. There is a lot of work needed to make it ready for real usage.
> 
> I've been working on #129 to introduce raster images as an alternative to vectors. Raster images have the advantage of improved frame rate (less jank) and they only need to be rendered the first time since they can be stored in the cache for future use. The result is quite good. Of course, raster images aren't as sharp at fractional zoom levels. #129 is now available as version 3.3.0 of this library.

Thank you, I have already tried this update, it work really faster, but I still have freezes ((. In any case any calculation in main thread entail freezes of the UI.  Even if they won't be visible for the human eye, but they will  still affect any timer or animation in app.
I hope this problem will be solved with impeller, and there will be a mechanism for rendering in non main thread.

**Comment by JaffaKetchup** [comment]
@greensopinion Please note https://pub.dev/packages/flutter_map/versions/4.0.0-dev.1/changelog

**Comment by greensopinion** [comment]
Thanks Luke. Is the update to let me know about API changes or does it
specifically relate to performance?

**Comment by JaffaKetchup** [comment]
@greensopinion Both. Probably not a massive performance improvement, but there is a chance it will help. There are major API changes - but I'm not sure if any of them will affect you. Probably the most major for plugin maintainers is the renaming of `Coords` to `TileCoordinates`.

**Comment by zyhong** [comment]
Maybe this rive render could help the performance issue? 
https://github.com/rive-app/rive-flutter
https://www.reddit.com/r/FlutterDev/comments/137s0ov/benchmarking_rive_renderer_vs_skia_vs_impeller/?rdt=57000

**Comment by mdmm13** [comment]
Hey @greensopinion, was trying to follow up on your [discussion here](https://github.com/rmawatson/flutter_isolate/issues/128#issuecomment-1601887798) about using [flutter_isolate](flutter_isolate) to offload processing yet still able to call dart.ui functions outside the main isolate. It's not the full solution I understand from [your comment here](https://github.com/greensopinion/flutter-vector-map-tiles/issues/93#issuecomment-1268617838), but part of it. The other part being the complexity of the theme, which you're trying to solve by selectively delaying the rendering if I understood correctly.

I'm not sure I got it right, but per nmfisher's (flutter_isolate maintainer) last comment [here](https://github.com/rmawatson/flutter_isolate/issues/138#issuecomment-1702005842), the status on [this PR](https://dart-review.googlesource.com/c/sdk/+/292320) being merged into dart's main branch and the [comment here](https://github.com/flutter/flutter/issues/10647#issuecomment-1704222685), it appears that the functionality that Impeller previously blocked on iOS should now be back again with Dart 3?

Does this mean, flutter_isolate + selective delays COULD solve many of our performance issues? What am I missing?

**Comment by mdmm13** [comment]
> Hey, @greensopinion - have you considered drawVerticesRaw for rendering? It seems to be orders of magnitude faster than using widgets or canvas. This video explains it very thoroughly and the bookmarked section shows a comparison between the different methods; https://youtu.be/pD38Yyz7N2E?t=742

Would probably solve all problems with vector maps without going GL?

**Comment by JaffaKetchup** [comment]
Just for context, we're looking into potentially using `drawVertices` to draw polygons in FM, but that's still in mostly private discussion.
You may also be interested in dart_earcut (self-promo :D), which will perform the triangulation needed for `drawVertices` extremely quickly (it doesn't work well on self-intersecting polys, they'll need to be simplified into multiple simple polys).

**Comment by greensopinion** [comment]
Looks promising! Feel free to prototype something.

**Comment by DavidDohmen** [comment]
Unfortunately I have neither the time nor skillset to do so. We would however contribute financially with a bounty, if a certain speed level is reached, because then flutter_map would become feasible as an alternative to mapbox for us. 
So if anyone with the right skillset wants to pick this up, let's talk!

**Comment by JaffaKetchup** [comment]
@j-bbr Right now we are not implementing `drawVertices`, and if we do, it'll be for `Polygon`s only: we have no plans to make a vector layer for now.

**Comment by greensopinion** [comment]
Go for it!

**Comment by j-bbr** [comment]
@greensopinion Just to keep an open mind for which improvements to go for I was wondering whether I could get your opinion on the following so I don't go down a path that has already turned out to be a dead-end for you:
Besides drawing faster, potentially with `drawVertices` there is of course also the option to draw slower :) meaning in this case to spread drawing out over multiple frames. You're of course already doing that with Text but maybe this could be extended to partial drawing of layers (at the moment all layers except symbols are drawn in one frame, please correct if I am wrong here)?

Rough Todo-List:
 keeping track of partially drawn tiles
 partially drawing layers (fixed number initially, maybe within a time budget later)
 retaining the partially drawn state efficiently for continuation in the next frame (not possible afaik with `CustomPainter`, maybe the `VectorTileLayerMode.raster ` code can be a start for this, or triangle lists for drawVertices)

Any thoughts, pitfalls you can already see here?

**Comment by greensopinion** [comment]
I have a tough time imagining a good user experience if maps have features
appearing/disappearing while zooming or panning.
I'd focus on two approaches:

1) make what we do today faster, and

2) experiment with raster base layer tiles with vector detail

Raster tiles could be rendered using vector data on an isolate (eg with
flutter_isolate) without text or other symbols. This would offload that
work from the UI thread.

Before deciding on either it's probably worthwhile looking at the profiler
to see what's expensive. For example, if polygons aren't contributing much
to jank then moving those off to a raster base layer might not improve
anything. maybe you'll have new/better ideas after checking the profiler
since right now I'm just guessing.

Have fun!

**Comment by greensopinion** [comment]
I've pushed a prototype rendering raster tiles from vector tiles. To use it, you can check out the following branches:

* https://github.com/greensopinion/flutter-vector-map-tiles/tree/raster-on-isolate
* https://github.com/greensopinion/flutter-vector-map-tiles-examples/tree/raster-on-isolate

This prototype is pretty rough, with notable
deficiencies:

- isolates are leaked whenever the tile provider
  is recreated
- data transfer between isolates causes each tile
  to incur about 2s of latency, which in turn
  causes queueing
- queueing causes additive latency
- conversion of raster images to png occurs twice

Switching between raster/vector and panning provides a good contrast in jank and tile latency.

**Comment by Soap-141** [comment]
I don't know if it would be useful for you guys but I tried Here SDK for Flutter and the map using vector doesn't have any performance issue even in 3D. It's insane and I wonder how they are doing it.

**Comment by JaffaKetchup** [comment]
@Soap-141 They're probably using a platform view ~which is cheating :D~. But on a serious note, FM is 100% Dart, which means maximum flexibility and stability, but potentially sometimes at the cost of performance.

**Comment by mdmm13** [comment]
> I don't know if it would be useful for you guys but I tried Here SDK for Flutter and the map using vectors doesn't have any performance issue even in 3D. It's insane and I wonder how they are doing it.

They're using native SDKs with a Flutter lay on top AFAIK (https://www.here.com/docs/bundle/sdk-for-flutter-explore-api-reference/page/core.engine/SDKNativeEngine-class.html). You'll find plenty of references to JNI etc. in their changelogs.

**Comment by JaffaKetchup** [comment]
Just an update, we are now working on `drawVertices` for FM Polygons, and there's a working implementation for non self-intersecting polys (holes are supported). It's drastically improved raster times, and uses dart_earcut for triangulation. Maybe it could be helpful for those wanting to see how to do it here.
There's an open PR by me in FM.

**Comment by greensopinion** [comment]
That's cool Luka. Let us know how it goes.

**Comment by ignatz** [comment]
Hey folks, just found this thread. Inspired by https://github.com/fleaflet/flutter_map/pull/1800, I took a quick stab at performance improvements:

 1. I did a bunch of micro-optimizations. Most are probably not even worth.
 2. Batched the canvas draws
 3. (I also made web rendering work)

As a result (mostly of (2)) my raster times more than halved, at least for simple pans (my setup: linux notebook and compiled in profile mode and concurrency set to 0 to make this as straightforward as possible):

vanilla:
https://github.com/greensopinion/flutter-vector-map-tiles/assets/111986/92d08665-3d15-4104-ba37-2a4f5dc0886d

optimized:
https://github.com/greensopinion/flutter-vector-map-tiles/assets/111986/bfded73d-bea0-4111-9470-c778e539cb12

Depending how I'm moving the map I'm still seeing quite a bit of (presumably) shader jank. Hopefully impeller will help with that.

I'm also hopeful that using `drawVertices` would improve raster performance even further.

**Comment by ignatz** [comment]
I also quickly switched to drawVertices, it does help (avg raster times drop from 3.5ms to 3ms) but most of the raster time is spent on lines:


https://github.com/greensopinion/flutter-vector-map-tiles/assets/111986/ed16be9d-f2cd-41c3-87b3-2b2c32faab16

**Comment by josxha** [comment]
As far as I've seen the line renderer currently uses `drawPath()`. Do you think `drawPoints()` or event `drawRawPoints` with `PointMode.lines` could help here?

**Comment by ignatz** [comment]
> As far as I've seen the line renderer currently uses `drawPath()`. Do you think `drawPoints()` or event `drawRawPoints` with `PointMode.lines` could help here?

I've experimented with both for the polygon layer, so I'd naively say no. Then again, based on my late edit, I'm not even sure that lines are that dominant. I'd expect geometry simplifications for both lines and files to probably help more.

**Comment by ignatz** [comment]
![Screenshot_20240202-183422_vector_map_tiles_example](https://github.com/greensopinion/flutter-vector-map-tiles/assets/111986/4f117f19-a10d-4b00-a7a4-18ae3c5c3953)
![Screenshot_20240202-183335_vector_map_tiles_example](https://github.com/greensopinion/flutter-vector-map-tiles/assets/111986/6f9b0c21-cddd-4e43-a26e-dc1be639a263)

I quickly tried Impeller on Android. The result was shockingly bad. I hope this is to blame on the alpha status. Maybe someone with an iOS device can try?

**Comment by mdmm13** [comment]
Possibly related, but the Flutter issues are full of performance problems as of 3.16 (e.g. [here](https://github.com/flutter/flutter/issues/140257), [here](https://github.com/flutter/flutter/issues/141961)), all related to vector graphics. They concluded that it's [caused by tessellation](https://github.com/flutter/flutter/issues/140257#issuecomment-1888336577).

What does your CPU profile look like? Same cause? 

One interesting aspect is that there appears to be a limit on the number of [canvas.drawX operations of ~100 per frame](https://github.com/flutter/flutter/issues/138004). Anything above that and it degrades massively. Sounds like our issue?

**Comment by mdmm13** [comment]
And [here](https://github.com/flutter/flutter/issues/131345) you've got it white on black that drawVertices took a hit with Impeller. The popular [YouTube video](https://www.youtube.com/watch?v=pD38Yyz7N2E) we used to come up with the drawVertices idea is from Jan 2023. It was not till May 2023 that iOS made Impeller the default in iOS.

Thoughts?

**Comment by ignatz** [comment]
> That's interesting @ignatz. I haven't run it on this plugin yet, but I can try soon hopefully (would be grateful if you could post a link to the implementation you're running above).

Happy to share my code but it's a messy hodgepodge with inter-project dependencies. Meanwhile, I'd naively expect the effect to be the same independent of drawVertices and batching.

> I would've expected Impeller to make a big difference, especially to lines. It certainly does for the polygon layer on my Android device - can you test that on your device as well? Does your device have Vulkan support?

My device is pretty ancient but should support vulkan. I can try the polygon layer in due time as well but so far I wasn't too interested since I don't feel to keen chase down buggy alpha-quality software.

> What does your CPU profile look like? Same cause?

I haven't profiled the raster thread yet. I only gave it a quick spin after the raving reviews by Jaffa and feeling a bit too cocky after lowering the raster times by 70% already. As I said, I'm not particularly keen on chasing down bugs in alpha quality software. For the time being, I feel my efforts would be spent better somewhere else, e.g. geometry simplifications.

Tessellation or not, the issue you linked sounds very relevant and plausible: https://github.com/flutter/flutter/issues/131345 . If that was the case, we're not talking bugs but rather crappy implementation. According to the issue, a solution is on halt for "higher priority" work. (tangential rant: gotta love flutter government, hope they'll eventually hand it over to a foundation). Meanwhile, playing with the blending might provide some insights :woman_shrugging:

**Comment by JaffaKetchup** [comment]
See https://github.com/flutter/flutter/issues/131345#issuecomment-1925338105 for some performance comparisons.

**Comment by ignatz** [comment]
I spent a bit more time doing some more draw batching. For the same basic panning example my raster times are now down to 2.4ms on average compared to 9.4ms originally.  That's roughly a 75% reduction, without any geometry simplifications.

https://github.com/greensopinion/flutter-vector-map-tiles/assets/111986/c7536de1-1907-44b2-90df-cf9938b18d7c

Web is still a bit janky for the same scene at <20FPS but since I'm getting almost 60 FPS for scenes with less geometry in them, I'm fairly optimistic that simplifications might even make web viable :woman_shrugging: 

I did clean up my changes a bit, in case folks would like to join: https://github.com/ignatz/dart-vector-tile-renderer/commits/draw_less_smile_more/

**Comment by greensopinion** [comment]
Hey all, great to see the collaboration here. I haven't had a chance to catch up on all of the comments but it looks like some promising progress. 

If anyone's interested in helping to move things along, #196 is in need of a review.

**Comment by ignatz** [comment]
> If anyone's interested in helping to move things along, https://github.com/greensopinion/flutter-vector-map-tiles/pull/196 is in need of a review.

I helped. I closed it :) . 

My current thinking, after speeding up https://github.com/greensopinion/dart-vector-tile-renderer, is that I'll pursue tile lifecycle management different from https://github.com/greensopinion/flutter-vector-map-tiles.git.

**Comment by ignatz** [comment]
I spend a little bit of time on writing my own vector tile layer.

Disclaimer: 
 * It has a few bugs still, e.g. label placement can be offset or labels can be clipped.
 * There's many things I could be smarter about

That said, it renders at 40+ FPS in the browser

https://github.com/greensopinion/flutter-vector-map-tiles/assets/111986/e8f18648-bfc6-4ec1-bc33-b64671f7ba24

**Comment by ibrierley** [comment]
One thing I'd be interested in seeing is how a city eg London is with zooming in/out these days, as that was always the area that made me hesitant with performance. Pans will normally be optimised by flutter I think, as it doesn't need to recalculate linecaps etc, but when zooming in/out I think they do (I may be a bit out of touch on that tho).

**Comment by ignatz** [comment]
To expand a bit on what you're saying:

  * It gets slow whenever you re-render your geometry
  * And it gets even slower when there's a lot of geometry (like Londong or the Alps in the video above).

By default whenever you use a CustomPaint in flutter, flutter will rerender when the parent rebuilds because it doesn't know. You can add a RepaintBoundary and implement `shouldRepaint` to give Flutter a helping hand (sometimes it will ignore you) on deciding on when to re-render.

What you're saying about pans is correct to the extent that we all suck at math enough that we're afraid of `Transformation`. We typically only use `Positioned` in combination with RepaintBoundaries. Consequently, pans tend to be cheap, while zooming and rotating tends to be expensive triggering re-renders.

When I mentioned tile lifecycles, I really just meant we need to hold on to the tiles and simply not re-render them. I'm also lazy so I didn't really write my own tiling layer. What you're seeing is FlutterMap's default tiling layer with a TileProvider that renders vector tiles into images with an extra Symbol/Text only layer on top. Just like in David's implementation the cost of the labels is simply mitigated by taking them off screen while rotating and zooming. When they finally pop in, the rendering janks terribly but you just don't see it on an otherwise static screen.

Long story short, London would affect the initial image render times but afterwards you're just moving around the same static image.

**Comment by greensopinion** [comment]
That's cool Sebastian. We should learn from what you did. How is your
lawyer different than what is already provided by this library?

**Comment by mdmm13** [comment]
> I spend a little bit of time on writing my own vector tile layer.
> 
> Disclaimer:
> 
> * It has a few bugs still, e.g. label placement can be offset or labels can be clipped.
> * There's many things I could be smarter about
> 
> That said, it renders at ~50 FPS in the browser
> 
>  Screencast_webrenderer.mp4

@ignatz - Is that the "draw_less_smile_more"  branch and/or "cache_n_web"? We've been tinkering with vector maps for years, but never made the jump due to performance on Android/ Web, iOS was always OKish. Would love to try this out.

David - I recall a remark somewhere (can't find the quote just now) that spreading out rendering of layers over multiple frames would result in elements showing up after a little delay and that this might look weird - https://maps.earth appears to do just that and IMHO it looks totally fine if that's the price to pay for having Android/ web performance. Take a look and see how it feels to you.

**Comment by ignatz** [comment]
> @ignatz - Is that the "draw_less_smile_more" branch and/or "cache_n_web"? We've been tinkering with vector maps for years, but never made the jump due to performance on Android/ Web, iOS was always OKish. Would love to try this out.

Yes an no. The "draw_less_smile_more" branch has the 75% raster time reduction (at least pretty much all of it, I have more private branches trying out things). If you use that branch with the upstream tile layer, that will get you some way.

The quoted 50FPS on Web are after taking these improvements and writing a new tile layer that clings on to tiles. I haven't shared the latter yet. I'm contemplating what I should do with it, do I wanna maintain this, should I actually switch to maplibre-rs for rendering :woman_shrugging: 

> David - I recall a remark somewhere (can't find the quote just now) that spreading out rendering of layers over multiple frames would result in elements showing up after a little delay and that this might look weird - https://maps.earth/ appears to do just that and IMHO it looks totally fine if that's the price to pay for having Android/ web performance. Take a look and see how it feels to you.

W/o the quote it's hard to say anything of substance other than to point out the fact that this is already the case. After all your downloading the tiles independently and they'll show up whenever your done loading, preprocessing, and rendering. Are you referring to the case when the tile is already locally cached?

**Comment by mdmm13** [comment]
> The quoted 50FPS on Web are after taking these improvements and writing a new tile layer that clings on to tiles. I haven't shared the latter yet. I'm contemplating what I should do with it, do I wanna maintain this, should I actually switch to maplibre-rs for rendering 🤷‍♀️

All other Flutter vector maps use the GPU, either via native, webview, etc. As you stated [here](https://github.com/fleaflet/flutter_map/pull/1800#issuecomment-1924038470) though, it'll be a while until we get GPU access through Flutter. maplibre-rs would be fantastic, but the WebGPU standard is still being developed and WEB performance there is also not great (yet) and it's going to be a while before that's an option. Maybe the answer is that pure Flutter vector performance will only catch up in 2025 (?).

> I'm pretty sure that with some investment the vector performance could be stellar. I don't see why it couldn't. Web clearly demonstrates that vector renders are viable.

After your experiments, do you think there's a play here to get vector to perform without GPU?

> W/o the quote it's hard to say anything of substance other than to point out the fact that this is already the case.

Found and added the quote above. It was about actively spreading out the rendering of different layers to improve performance.

**Comment by ignatz** [comment]
Let me unpack, you have a lot of points where I'm not always sure if or how they relate.

> All other Flutter vector maps use the GPU, either via native, webview, etc. 

You mean mapbox-gl and maplibre-flutter? I've only briefly looked into their implementations. Does anyone do CPU-only rendering?

> As you stated https://github.com/fleaflet/flutter_map/pull/1800#issuecomment-1924038470 though, it'll be a while until we get GPU access through Flutter.

Right, importantly though 'low level' GPU access. When using flutter, depending on your engine backend, you're certainly using the GPU.

> maplibre-rs would be fantastic, but the WebGPU standard is still being developed and WEB performance there is also not great (yet) and it's going to be a while before that's an option.

I can't comment on the web performance of maplibre-rs, the desktop performance seemed fine. I'm not sure why you're bringing up WebGPU. I assume that maplibre-rs would use WebGL on web. It has a variety of backends. It presumably used vulkan or OpenGl when running on my desktop: https://github.com/maplibre/maplibre-rs/blob/095d5fbbfb0fea4bfa680e9534944813447b220e/README.md?plain=1#L57 . Is there any concern I'm missing?

> Maybe the answer is that pure Flutter vector performance will only catch up in 2025 (?).

Catch up to what? Is your statement regarding canvas-based rendering and the respective cross-platform backends using skia (wich has both a OpenGl and Vulkan beackend), impeller (which is Vulkan only?), ... ?

> After your experiments, do you think there's a play here to get vector to perform without GPU?

I think there's a play with not rendering and frequently re-rendering vector tiles synchronously on the raster thread's frame generation path.

Regarding with or w/o GPU, when you use `Picture.toImage` to rasterize a Picture (canvas + ops) what makes you think you'd not be using the GPU? Ultimately, it depends on the flutter engine's implementation. I can't tell you for sure if and when it would fall back to a purely CPU-based rasterizer. Skimming the implementations, impeller seems to always be using the GPU, the Skia backend is a lot for conditional: https://github.com/flutter/engine/blob/6d524588c0f9a4c73142f8659e8f113e6e6492ce/lib/ui/painting/picture.cc#L42 (it looks like i may depend on thread locals, i.e. which thread you're running on, etc)

> Found and added the quote above. It was about actively spreading out the rendering of different layers to improve performance.

Ok, David's quote is:  "I have a tough time imagining a good user experience if maps have features appearing/disappearing while zooming or panning."

I agree 100% with David. To be clear, https://maps.earth/  does not do that either. When a tile is loaded and you zoom a little the tile stays on screen. Imagine you zoom and the already loaded tile would vanish to then come back a few frames later.

**Comment by mdmm13** [comment]
> Let me unpack, you have a lot of points where I'm not always sure if or how they relate.

When I say "GPU", I really mean native and what I presume is better GPU access to explain the massive performance differences, apologies for that.


> > All other Flutter vector maps use the GPU, either via native, webview, etc.
> 
> You mean mapbox-gl and maplibre-flutter? I've only briefly looked into their implementations. Does anyone do CPU-only rendering?

Yes, I do mean flutter-maplibre-gl (layer on top of native/gl_js), HERE (layer on top of native), Google Maps, ... all the vector ones that somehow perform well. Haven't looked at their native implementations though.


> > As you stated [fleaflet/flutter_map#1800 (comment)](https://github.com/fleaflet/flutter_map/pull/1800#issuecomment-1924038470) though, it'll be a while until we get GPU access through Flutter.
> 
> Right, importantly though 'low level' GPU access. When using flutter, depending on your engine backend, you're certainly using the GPU.

Yes, of course :)


> > maplibre-rs would be fantastic, but the WebGPU standard is still being developed and WEB performance there is also not great (yet) and it's going to be a while before that's an option.
> 
> I can't comment on the web performance of maplibre-rs, the desktop performance seemed fine. I'm not sure why you're bringing up WebGPU. I assume that maplibre-rs would use WebGL on web. It has a variety of backends. It presumably used vulkan or OpenGl when running on my desktop: https://github.com/maplibre/maplibre-rs/blob/095d5fbbfb0fea4bfa680e9534944813447b220e/README.md?plain=1#L57 . Is there any concern I'm missing?

I thought you were talking about the maplibre-rs PoC (from the [maplibre-rs Github](https://github.com/maplibre/maplibre-rs#description): "maplibre-rs is a portable and performant vector maps renderer. We aim to support web, mobile and desktop applications. This is achieved by the novel [WebGPU](https://www.w3.org/TR/webgpu/) specification."). No concerns if you're going for the GL ones.


> > Maybe the answer is that pure Flutter vector performance will only catch up in 2025 (?).
> 
> Catch up to what? Is your statement regarding canvas-based rendering and the respective cross-platform backends using skia (wich has both a OpenGl and Vulkan beackend), impeller (which is Vulkan only?), ... ?

I'm referring to the other vector maps that perform surprisingly well in Flutter. That's the key question I believe most people are trying to answer when choosing between flutter_map + vector plugin and the ones mentioned above.

 
> > After your experiments, do you think there's a play here to get vector to perform without GPU?
> 
> I think there's a play with not rendering and frequently re-rendering vector tiles synchronously on the raster thread's frame generation path.

Got it, thank you.


> Regarding with or w/o GPU, when you use `Picture.toImage` to rasterize a Picture (canvas + ops) what makes you think you'd not be using the GPU? Ultimately, it depends on the flutter engine's implementation. I can't tell you for sure if and when it would fall back to a purely CPU-based rasterizer if ever. Skimming the implementations, impeller seems to always be using the GPU, the Skia backend is a lot more conditional: https://github.com/flutter/engine/blob/6d524588c0f9a4c73142f8659e8f113e6e6492ce/lib/ui/painting/picture.cc#L42 (it looks like i may depend on thread locals, i.e. which thread you're running on, etc)

As per my original clarification above. I'm aware Flutter uses the GPU :)

> Ok, David's quote is: "I have a tough time imagining a good user experience if maps have features appearing/disappearing while zooming or panning."
> 
> I agree 100% with David. To be clear, https://maps.earth/ does not do that either. When a tile is loaded and you zoom a little the tile stays on screen. Imagine you zoom and the already loaded tile would vanish to then come back a few frames later. You might be thinking about the initial first render. Sure that can happen whenever and that's already the case.

Alright, then I misunderstood. From the various Flutter links I shared, to me the logic was: (A) we've got tons of drawX calls beyond the [100](https://github.com/flutter/flutter/issues/138004) or so that 1 post mentioned as a performance drop-off point (B) with vector we can choose to delay rendering layers until others are done (C) I'd personally rather have performance and something looking a bit "weird" with layers like text and polygons if that was a trade-off that was on the table.

**Comment by greensopinion** [comment]
@ignatz 

> I spent a bit more time doing some more draw batching. ...
> I did clean up my changes a bit, in case folks would like to join: https://github.com/ignatz/dart-vector-tile-renderer/commits/draw_less_smile_more/

I took a quick look over your changes, they look really good. A few questions:

- do you have metrics on batching, i.e. the average batch size?
- do you have a repeatable method for measuring improvements?
- have you looked at whether transparency (e.g. in the line color) works well with drawVertices?
- how is your "clings on to tiles" layer different from using [raster mode](https://github.com/greensopinion/flutter-vector-map-tiles/blob/main/lib/src/vector_tile_layer_mode.dart#L8)?

**Comment by greensopinion** [comment]
Thanks for the details! I'm still confused as to how your cling technique
is different than what we are already doing in this library:
https://github.com/greensopinion/flutter-vector-map-tiles/blob/main/lib/src/grid/grid_layer.dart#L144

**Comment by ignatz** [comment]
Maybe you're expecting too much :). I'm rendering into an Image because that let's me control the lifecycle. You're drawing into a CustomPaint widget. RepaintBoundary gives one some limited control, which you're using to prevent re-renders when panning. However when you're zooming or rotating, you're re-rasterizing on every frame

**Comment by greensopinion** [comment]
Sounds like we are doing the same thing, since with raster mode we are
rendering to an image too.

**Comment by ignatz** [comment]
Shame on me, digging through all the layers I actually missed that switch. I took a quick look and unsurprisingly the performance in raster mode is much much better.
In that case the only difference is that I'm rendering labels and tiles in separate passes, i.e. the labels (un)rotate independent from the tiles.

**Comment by greensopinion** [comment]
Glad we both had the same good idea!

**Comment by ignatz** [comment]
Agreed. I also never wanted to imply any of it is rocket science. I did it because tapping into the existing FlutterMap tilelayer infra I could do it with rather little effort.
FWIW, I do think that the hybrid approach helps with selling the "illusion" or at least reclaiming some of the vector map benefits, see:


https://github.com/greensopinion/flutter-vector-map-tiles/assets/111986/ba168ea0-ea0a-4919-8f39-8a7b3f04cd49

**Comment by JaffaKetchup** [comment]
I have just had a little experiment with a custom Mapbox style, and noticed that setting `VectorTileLayerMode.vector` actually gave considerably better performance on my computer than the raster mode. It may be worth revising the documentation.

**Comment by greensopinion** [comment]
That's cool @JaffaKetchup - can you quantify it?

**Comment by j-bbr** [comment]
I've put a profiling / integration test for vector maps based on this example from the [documentation](https://docs.flutter.dev/cookbook/testing/integration/profiling) in this repository [here](https://github.com/j-bbr/vector_tiles_performance_profiling). The test simulates map interaction (moving the map, zooming in, flinging the map again, zooming out etc) and records performance related data in a summary (examples attached) and an event trace .json file. The trace file can be viewed with chrome://tracing (see image)
<img width="2345" alt="ChromeProfiling" src="https://github.com/greensopinion/flutter-vector-map-tiles/assets/3635497/b416de14-033f-4a40-b9de-c25762b84b36"> It should be straightforward to change the map vector style, map starting location, vector vs. raster mode,  package version (released or from git branch / revision ) and get reproducible, comparable results. 

As an example I've attached 2 summaries, profiled on a Samsung S20, using the Stadia Maps OSM Bright style once in vector and once in raster mode and moving / zooming over Tokio.
[tokio_map_timeline.timeline_summary_raster.json](https://github.com/greensopinion/flutter-vector-map-tiles/files/14343741/tokio_map_timeline.timeline_summary_raster.json)
[tokio_map_timeline.timeline_summary_vector.json](https://github.com/greensopinion/flutter-vector-map-tiles/files/14343742/tokio_map_timeline.timeline_summary_vector.json)

**Comment by JaffaKetchup** [comment]
@j-bbr It would be awesome if you could make a version without the vector tiles stuff, and it would be much appreciated as a test in FM core!

**Comment by mdmm13** [comment]
Just FYI, interesting [debouncing support](https://github.com/fleaflet/flutter_map/pull/1840) being added to the core `flutter_map`, to drive performance for lower-end devices and web. Great video showing impact [here](https://github.com/fleaflet/flutter_map/assets/69913791/ba2251cd-999f-44c2-8e3f-efde3ede40d5). 

Might be worth checking out for vector also, since the best performance optimization is doing less computations in the first place.

**Comment by Hellomik2002** [comment]
I am not sure but maybe it can be helpful?  https://pub.dev/packages/dart_ui_isolate
@chipweinberger Released new plugin

**Comment by JaffaKetchup** [comment]
> I'm curious as to why the default debounce is 0ms.

We decided 0 was best as it the feature does impact functionality a fair amount. New tiles aren't loaded at all until the debounce is complete: a long fling will not load any tiles, for examples.

**Comment by greensopinion** [comment]
> I am not sure but maybe it can be helpful? https://pub.dev/packages/dart_ui_isolate @chipweinberger Released new plugin

@Hellomik2002 cool to see more options come up. Do you have any idea how this would be an improvement over `flutter_isolate`?

We've [prototyped something with `flutter_isolate`](https://github.com/greensopinion/flutter-vector-map-tiles/issues/120#issuecomment-1867984099) which makes interactions glassy smooth - but it needs work before it could be useful in a real app. Feel free to try it out replacing `flutter_isolate` with `dart_ui_isolate`.

**Comment by ignatz** [comment]
Since dart_ui_isolates came up, I'm going to share my vector tile layer: https://github.com/ignatz/vector_tile_it/tree/main/example .

It contains all the optimizations mentioned before and is just generally a bit less layered. That said, it's pretty untested and I'm sure that my wild experiments introduced some regressions. I'm not using it productively myself since I'm still contemplating switching to a rust based rasterizer (at least for the tiles).

Enough disclaimer, I played with flutter_ui_isolate to see how much of an impact it has. That said, it's important to note that flutter_ui_isolates are only supported on android, ios and osx. So if you wanna play around with it, it can be easily disabled.

> Do you have any idea how this would be an improvement over flutter_isolate?

The two main differences are that they have less memory overhead and you can use dart:ui (https://github.com/flutter/flutter/issues/10647). Based on the issue I would have naively expected you to encounter issues but maybe things have improved since the bug was filed in 2017 :woman_shrugging:

**Comment by mdmm13** [comment]
> Thanks @mdmm13 that's helpful! I'm curious as to why the default debounce is 0ms.
> 
> Would you be interested in submitting a PR? A good place to start: https://github.com/greensopinion/flutter-vector-map-tiles/blob/main/lib/src/grid/grid_layer.dart#L144

Sure, but won't it automatically be added as soon as `flutter_map` gets released as v6.1 or v7? It's already been merged into master there?

**Comment by Soap-141** [comment]
Someone know when it will be available?

**Comment by greensopinion** [comment]
> Since dart_ui_isolates came up, I'm going to share my vector tile layer: https://github.com/ignatz/vector_tile_it/tree/main/example .

@ignatz very cool to see how you were able to integrate the renderer with very little code. `vector_map_tiles` suffers from code complexity due to the many iterations it took to evolve and my many hacks. If you've learned anything that you would like to see make it back into `vector_map_tiles` let me know.

**Comment by ignatz** [comment]
I'm not quite sure what you have in mind. Don't get me wrong, I'm super appreciative of the work that you've done especially for `vector_tile_renderer` but I wrote my own tile layer on top because I didn't think it was possible for me to make the changes I wanted to make to `vector_map_tiles`. 

You can certainly pull in my changes to `vector_tile_renderer`, they should benefit any user and were some -70% reduction for Mapbox terrain maps (I don't even remember anymore).

As for the tile layers, I'm not sure what "I would like to see make it back" or how that would work. My layer takes the hybrid approach: rasterized terrain with crisp vector labels and symbols, has pretty much all I'd expect from a tile layer, and is ~1/3 the code. At the expense of sounding accidentally presumptuous, if you're not happy with the state of `vector_map_tiles`, I'd love for you to take ownership in `vector_tile_it` (the name is negotiable :) )

Also to be clear, my initial excitement has worn off a bit. I'm not sure that even with all these improvements and changes, flutter-canvas-based vector tiles would meet my personal performance goals. As I've mentioned before, I've toyed with the idea of pushing the rasterization to a mapboxlibre-rs renderer :woman_shrugging:. But even then it wouldn't be true vector maps but rather faster raster maps with vector labels. I certainly don't get the point of rasterizing everything, why wouldn't I just use raster map tiles? (I guess if you really wanted to change the styling on the client but that sounds rather niche)

**Comment by mdmm13** [comment]
> Tessellation or not, the issue you linked sounds very relevant and plausible: [flutter/flutter#131345](https://github.com/flutter/flutter/issues/131345) . If that was the case, we're not talking bugs but rather crappy implementation. According to the issue, a solution is on halt for "higher priority" work. (tangential rant: gotta love flutter government, hope they'll eventually hand it over to a foundation). Meanwhile, playing with the blending might provide some insights 🤷‍♀️

Just FYI, Flutter 3.22 appears to improve [tessellation issues drastically](https://medium.com/flutter/whats-new-in-flutter-3-22-fbde6c164fe3). Might help.

**Comment by ignatz** [comment]
I just quickly ran 3.19 and 3.22 side-by-side. I wasn't sure if Linux was also using the new vulkan-based impeller renderer, looking at the output I'm leaning yes:

![Screenshot from 2024-05-15 10-42-09](https://github.com/greensopinion/flutter-vector-map-tiles/assets/111986/3da97670-1c89-4754-a42f-414f4cce27b9)

Maybe worth waiting a few fixes :hide:

Also worth highlighting from the article you linked:

> While pleased with these improvements, there is still more work to do. Among other opportunities, we are aware that polyline generation remains prominent in CPU profiles, and we intend to investigate shifting this work to the GPU, as well.

**Comment by mdmm13** [comment]
> I just quickly ran 3.19 and 3.22 side-by-side. I wasn't sure if Linux was also using the new vulkan-based impeller renderer, looking at the output I'm leaning yes:

Thank you for the quick check - looks like a 20-30% performance improvement on average.

> Also worth highlighting from the article you linked:
> 
> > While pleased with these improvements, there is still more work to do. Among other opportunities, we are aware that polyline generation remains prominent in CPU profiles, and we intend to investigate shifting this work to the GPU, as well.

The fact that they're stating that in the official announcement means that they're probably working on it :-)

**Comment by Hellomik2002** [comment]
> I'm not quite sure what you have in mind. Don't get me wrong, I'm super appreciative of the work that you've done especially for `vector_tile_renderer` but I wrote my own tile layer on top because I didn't think it was possible for me to make the changes I wanted to make to `vector_map_tiles`.
> 
> You can certainly pull in my changes to `vector_tile_renderer`, they should benefit any user and were some -70% reduction for Mapbox terrain maps (I don't even remember anymore).
> 
> As for the tile layers, I'm not sure what "I would like to see make it back" or how that would work. My layer takes the hybrid approach: rasterized terrain with crisp vector labels and symbols, has pretty much all I'd expect from a tile layer, and is ~1/3 the code. At the expense of sounding accidentally presumptuous, if you're not happy with the state of `vector_map_tiles`, I'd love for you to take ownership in `vector_tile_it` (the name is negotiable :) )
> 
> Also to be clear, my initial excitement has worn off a bit. I'm not sure that even with all these improvements and changes, flutter-canvas-based vector tiles would meet my personal performance goals. As I've mentioned before, I've toyed with the idea of pushing the rasterization to a mapboxlibre-rs renderer 🤷‍♀️. But even then it wouldn't be true vector maps but rather faster raster maps with vector labels. I certainly don't get the point of rasterizing everything, why wouldn't I just use raster map tiles? (I guess if you really wanted to change the styling on the client but that sounds rather niche)

publish it, let's see it

**Comment by JaffaKetchup** [comment]
I was doing some experiments recently with [custom shaders](https://docs.flutter.dev/ui/design/graphics/fragment-shaders) in flutter_map to render raster tiles with opacity more efficiently. Unfortunately it wasn't useful in the end (barely any noticable performance improvement & https://github.com/flutter/flutter/issues/133944), but I did learn a tiny bit about shaders, so not all was wasted!

![Top secret discussion](https://github.com/greensopinion/flutter-vector-map-tiles/assets/58115698/7e4edc10-1448-45aa-b013-b21f7583c9fd)

(top secret behind the scenes chat ;))

However, I was wondering whether this would even be remotely possible here. I did some searching and found: http://kunzhou.net/zjugaps/pathrendering/ (https://web.archive.org/web/20240316141847/http://kunzhou.net/zjugaps/pathrendering/). And the screenshot on their front page... is a large OSM tile rendered with their renderer! They provide all the code and a paper, and seem to have both vertex and fragment shaders. I have no idea if the fragment shaders are enough without the vertex shaders (I have basically no idea what I'm doing with shaders), but they seem to have similar code that looks to do similar things, and are maybe directly compatible with Flutter (maybe with some adjustment?)?

**Comment by kekland** [comment]
I feel like the way to go would be to wait until https://github.com/flutter/flutter/blob/master/docs/engine/impeller/Flutter-GPU.md drops (so that there's no need to hack your way around trying to add OpenGL/Vulkan/Metal bindings on different platforms) to be able to use vertex shaders

**Comment by JaffaKetchup** [comment]
Yep, agreed, wasn't going to suggest implementing the bindings alone! (Hey, long time no see @kekland :D)

Of course, there's probably more that can be done here with Flutter. I haven't had a proper look through https://github.com/greensopinion/dart-vector-tile-renderer though. FM's polygon renderer is extremely complex now, and we've squeezed out everything possible, to the point where some of the biggest drains come from Flutter trying to render strokes (disabling caps and joins improves it somewhat, and using lines of 1 device pixels helps).  
In addition to the `drawVertices` mode (which sometimes helps, sometimes hinders), we use batching to join paths together, and minimize draw calls. This has a hugely positive effect on performance - although of course, it might be more complex to do this properly here, to preserve stacking order whilst keeping the order in such a way that batching is effective (we have to draw and prune the buffer whenever some characteristics between the last and current polygon change, and labels and opacity also need special expensive handling).

In summary, there's probably 3 things to try:

1. Improve the renderer using more advanced techniques - probably the easiest but still not very easy
2. Offload onto a UI-capable isolate/thread - easiest, but liable to break and introducing native code
3. Start experimenting with/wait for Flutter GPU support to use custom shaders

**Comment by kekland** [comment]
@JaffaKetchup Hi, yeah, great to see you :P

I'll try to play around with the existing GPU support whenever I'll have some spare time to see at what state it's currently at and whether we can benefit from parallelizing the process of converting the tile data to a vector. 

So far the performance seems to be alright when using the raster mode, but it would be great to have a full Google Maps-like experience. Haven't tested on lower-end devices yet though

**Comment by T-moz** [comment]
Wow, did you see the new announcement about flutter GPU preview ?
https://medium.com/flutter/getting-started-with-flutter-gpu-f33d497b7c11

**Comment by JaffaKetchup** [comment]
That's quite an announcement! I have no idea where to even begin implementing the algorithm I mentioned above, but it should at least now be possible!

(Strange, I was just checking the progress of flutter_gpu yesterday, wondering what was going on :D)

**Comment by mvarendorff2** [comment]
Hey thread!

I am currently looking into lags showing when moving the map while tiles are loading. [One particular comment](https://github.com/greensopinion/flutter-vector-map-tiles/issues/120#issuecomment-1975241249) stood out to me in this thread:

> We've https://github.com/greensopinion/flutter-vector-map-tiles/issues/120#issuecomment-1867984099 which makes interactions glassy smooth - but it needs work before it could be useful in a real app.

@greensopinion what work were you referring to there? Glassy smooth interactions sound exactly like what I am looking for at the moment and I'd love to contribute to bring using isolates along to a "production-ready" state!

I am also curious about thoughts from anyone here about `flutter_isolate` vs `dart_ui_isolate`. My understanding is that `dart_ui_isolate` would be preferable since we

a) don't need plugin support (and even if we did, since flutter 3.7 using plugins is supported in isolates)
b) do like the sound of instant isolates and 99% reduced RAM usage

but I might be missing something here. One thing that has me worried of sorts is the quiet state of the `dart_ui_isolate` repository. 

Let me know what you think! I will get to work on rebasing the isolates branch probably next week to see how performance changes in my little testbench and would very likely incorporate all other suggested required changes over the next week as well if possible.

**Comment by greensopinion** [comment]
Awesome, great to have someone looking into this. [the comment that you linked](https://github.com/greensopinion/flutter-vector-map-tiles/issues/120#issuecomment-1867984099) has 4 points that summarize the work that I thought was needed. 

My experiment isn't great code - just a hack to see what's feasible - so feel free to do something different. There is another implementation [described and linked in this thread](https://github.com/greensopinion/flutter-vector-map-tiles/issues/120#issuecomment-1977004234) that might be even better. 

I haven't looked into `dart_ui_isolate` vs `flutter_isolate`. I agree that we don't need plugin support, so potentially `dart_ui_isolate` is preferable. Something else to consider is long-term support and maintainability of the library.

Have fun!

**Comment by mvarendorff2** [comment]
Thanks a bunch!

A quick update from my end - I did rebase and play around with both vector_tile_it and the isolate-branch of this repository and hit much bigger surprises than anticipated in both implementations (and both dart_ui_isolate and flutter_isolate) including quite heavy visual artifacts:

![Screenshot 2024-11-12 153306](https://github.com/user-attachments/assets/180e00e1-51b0-4967-ae8e-f465ced48cd5)

and some tiles being rendered very evidently in the wrong place:

![grafik](https://github.com/user-attachments/assets/9bbd8769-73aa-49d5-a03d-79f2a5492c69)

Streets were also seemingly not rendered although I cannot tell with certainty whether that is related to styles or the rendering.

In case anyone is curious or wants to try it out themselves, the `styleUri` is `https://sgx.geodatenzentrum.de/gdz_basemapde_vektor/styles/bm_web_col.json`.

Since I was spearheading this advancement in the context of a client project with limited budget I unfortunately (and with a bad conscience) have decided with my PM to not pursue this route further for the time being 😔 

I really hope that I can get some spare time together to take another look at this maybe in december. This package has been invaluable for me in the past and I would love to squeeze a bit more performance out of it.

**Comment by greensopinion** [comment]
another try with `dart_ui_isolate`: #220

feel free to try it out and provide feedback

**Comment by nrbnlulu** [comment]
Might be out of scope though I wonder if its possible to use https://github.com/Maximkaaa/galileo as a rendering backend.

**Comment by giaur500** [comment]
Isn't it for Rust only?

**Comment by kekland** [comment]
@giaur500 it might be possible to interop this.

We've toyed around with straight up using maplibre native, bridging the C++ <-> Dart with ffigen, and rendering the map to a Texture. It works, although due to the threading model of Flutter it has its own issues. If it sounds interesting, the source is available in https://github.com/kekland/maplibre_flutter_map, although we've decided to not pursue this any longer.

There's also https://github.com/fzyzcjy/flutter_rust_bridge, which I haven't tried, but it seems to be solid.

If there's a way to integrate wgpu with a Texture widget, then it should theoretically be possible to do so.

**Comment by nrbnlulu** [comment]
Impeller doesn't support windows / linux yet. 
Also
1. it won't give you optimal performance due to well, dart.
2. writing a whole vector rendered from scratch just for flutter seems unreasonable to me.

We are currently investing in https://github.com/Maximkaaa/galileo (which might integrate with map-libre / mapbox).
in the future we will write an inter-op with flutter using the textures API.

**Comment by greensopinion** [comment]
For those interested in helping to move GPU support forward, there is some interesting work going on here: https://github.com/kekland/gpu_vector_tile_renderer

**Comment by Hellomik2002** [comment]
Yep We are doing this with Kekland, We are faced with some problems with shaders to make styles flexible and resolve memory issue on android. Also now Focued on symbols. We are little bit limited by Flutter GPU, it's not stable that's why gpu_vector_tile_renderer it is not ready yet. But with this 2 pull request we can be close to make something stable.
https://github.com/flutter/flutter/pull/168377
https://github.com/flutter/flutter/pull/168294
